persistence/sql/persister_oauth2.go

```diff
@@ -445,41 +445,62 @@ func toEventOptions(requester fosite.Requester) []trace.EventOption {
 	}
 }
 
-func (p *Persister) CreateRefreshTokenSession(ctx context.Context, signature string, requester fosite.Requester) (err error) {
+func (p *Persister) CreateRefreshTokenSession(ctx context.Context, signature string, accessTokenSignature string, requester fosite.Requester) (err error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.CreateRefreshTokenSession")
 	defer otelx.End(span, &err)
 	events.Trace(ctx, events.RefreshTokenIssued, toEventOptions(requester)...)
-	return p.createSession(ctx, signature, requester, sqlTableRefresh, requester.GetSession().GetExpiresAt(fosite.RefreshToken).UTC())
+
+	req, err := p.sqlSchemaFromRequest(ctx, signature, requester, sqlTableRefresh, requester.GetSession().GetExpiresAt(fosite.RefreshToken).UTC())
+	if err != nil {
+		return err
+	}
+
+	var sig sql.NullString
+	if len(accessTokenSignature) > 0 {
+		sig = sql.NullString{
+			Valid:  true,
+			String: x.SignatureHash(accessTokenSignature),
+		}
+	}
+
+	if err = sqlcon.HandleError(p.CreateWithNetwork(ctx, &OAuth2RefreshTable{
+		OAuth2RequestSQL:     *req,
+		AccessTokenSignature: sig,
+	})); errors.Is(err, sqlcon.ErrConcurrentUpdate) {
+		return fosite.ErrSerializationFailure.WithWrap(err)
+	} else if err != nil {
+		return err
+	}
+
+	return nil
 }
 
 func (p *Persister) GetRefreshTokenSession(ctx context.Context, signature string, session fosite.Session) (request fosite.Requester, err error) {
 	ctx, span := p.r.Tracer(ctx).Tracer().Start(ctx, "persistence.sql.GetRefreshTokenSession")
 	defer otelx.End(span, &err)
 
-	r := OAuth2RefreshTable{OAuth2RequestSQL: OAuth2RequestSQL{Table: sqlTableRefresh}}
-	err = p.QueryWithNetwork(ctx).Where("signature = ?", signature).First(&r)
-	if errors.Is(err, sql.ErrNoRows) {
+	var row OAuth2RefreshTable
+	if err := p.QueryWithNetwork(ctx).Where("signature = ?", signature).First(&row); errors.Is(err, sql.ErrNoRows) {
 		return nil, errorsx.WithStack(fosite.ErrNotFound)
 	} else if err != nil {
 		return nil, sqlcon.HandleError(err)
 	}
 
-	fositeRequest, err := r.toRequest(ctx, session, p)
-	if err != nil {
-		return nil, err
-	}
-
-	if r.Active {
-		return fositeRequest, nil
+	gracePeriod := p.r.Config().RefreshTokenRotationGracePeriod(ctx)
+	if row.Active {
+		// Token is active
+		return row.toRequest(ctx, session, p)
+	} else if row.FirstUsedAt.Valid &&
+		row.FirstUsedAt.Time.Add(gracePeriod).After(time.Now()) &&
+		gracePeriod > 0 {
```

Comment:
If the token was used, first used at is set, grace period is set, and grace period + last used has not yet passed, the token can be considered graceful.